Docs for CA2252 (#26149)
dotnet/docs: docs/fundamentals/code-analysis/quality-rules/ca2252.md

---
title: "CA2252: Opt in to Preview Features"
description: "This analyzer scans for preview API usage and produces a build error if an assembly has not opted into preview features"
ms.date: 09/14/2021
ms.topic: reference
f1_keywords:
- CA2252
- DetectPreviewFeaturesAnalyzer
helpviewer_keywords:
- DetectPreviewFeaturesAnalyzer
- CA2252
author: pgovind

> ms.author: pgovind

Reviewer: ```suggestion
ms.author: prgovi
```

dev_langs:
- CSharp

> - VB

Author: Actually this analyzer targets VB too. We should keep the VB entry around in that case right?

> - VB

Reviewer: This metadata is only used to define which languages the article has code snippets available in.

> - VB

Author: Ah makes sense. Removed now. Thanks!

---
# CA2252: Opt into Preview Features before using them

| | Value |
|-|-|
| **Rule ID** |CA2252|
| **Category** |[Microsoft.Usage](usage-warnings.md)|
| **Fix is breaking or non-breaking** |Non-breaking|

## Cause

A client uses preview APIs/types in their assembly without explicitly opting into it either locally or at the module/assembly level.

## Rule description

> When an API/Assembly marked with the `RequiresPreviewFeaturesAttribute` attribute is consumed, this analyzer checks if the call site has opted into preview features. A call site has opted into preview features if it is within the scope of a `RequiresPreviewFeaturesAttribute` annotation or if it is part of an assembly/module that has already opted into preview features. For ex:

Reviewer: ```suggestion
When an API or assembly that's decorated with the <xref:System.Runtime.Versioning.RequiresPreviewFeaturesAttribute> attribute is consumed, this rule checks if the call site has opted in to preview features. A call site has opted in to preview features if one of the following applies:

- It is within the scope of a `RequiresPreviewFeaturesAttribute` annotation.
- It is part of an assembly or module that has already opted in to preview features.

The following image shows an example of the CA2252 diagnostic.
```

> > ![CA2252 - Detect Preview Feature Usage](media/ca2252-basic.png)

Reviewer: ```suggestion
> ![Code editor with CA2252 warning.](media/ca2252-basic.png)
```

> Here, `Lib` is a preview type that is constructed in the `Main` method. `Main` itself is not annotated as a preview method, so diagnostics are produced on the 2 constructors calls inside `Main`.

Reviewer: ```suggestion
Here, `Lib` is a preview type that's constructed in the `Main` method. `Main` itself is not annotated as a preview method, so diagnostics are produced on the two constructor calls inside `Main`.
```

## How to fix violations

> There are 2 ways to fix violations:

Reviewer: ```suggestion
There are two ways to fix violations:
```

> 1. Bring a call site within the scope of an annotation by marking its parent with the `RequiresPreviewFeatures` annotation. In the example above, `APreviewMethod` is annotated with the `RequiresPreviewFeatures` attribute, so the analyzer ignores preview type usage inside `APreviewMethod`. It follows that callers of `APreviewMethod` will have to perform a similar exercise.

Reviewer: ```suggestion
- Bring a call site within the scope of an annotation by annotating its parent with `RequiresPreviewFeaturesAttribute`. In the previous example, `APreviewMethod` is annotated with the `RequiresPreviewFeatures` attribute, so the analyzer ignores preview type usage inside `APreviewMethod`. It follows that callers of `APreviewMethod` will have to perform a similar exercise.
```

> 

Reviewer: ```suggestion
```

> 2. If an assembly wishes, it can also opt into preview features at an assembly or module level. This indicates to the analyzer that preview type usage in the assembly is desired and, as a consequence, no errors will produced by this analyzer. This is the preferred way to consume preview dependencies. Setting the following property in a `.csproj` file will enable preview features inside the entire assembly:

Reviewer: ```suggestion
- You can also opt in to preview features at an assembly or module level. This indicates to the analyzer that preview type usage in the assembly is desired and, as a consequence, no errors will be produced by this rule. This is the preferred way to consume preview dependencies. To enable preview features inside the entire assembly, set the [EnablePreviewFeatures](../../../core/project-sdk/msbuild-props.md#EnablePreviewFeatures) property in a `.csproj` file:
```

> ```csharp

Reviewer: ```suggestion
  ```csharp
```

> 	<PropertyGroup>

Reviewer: ```suggestion
  <PropertyGroup>
```

> 		<EnablePreviewFeatures>true</EnablePreviewFeatures>

Reviewer: ```suggestion
    <EnablePreviewFeatures>true</EnablePreviewFeatures>
```

> 	</PropertyGroup>

Reviewer: ```suggestion
  </PropertyGroup>
```

> ```

Reviewer: ```suggestion
  ```
```

Author: Did you actually mean to delete the ``` here?

Reviewer: Hmm, this must be a GitHub bug. The actual suggestion is to indent the ``` by two spaces.

Reviewer 2: @gewarren Using backticks inside backticks don't always work well. Whenever I do suggestions involving backticks I use:

    ~~~suggestion
    suggestion containing ``` here
    ~~~

instead of

    ```suggestion
    suggestion containing ``` here
    ```

The following is using `~~~suggestion` instead of ``````````suggestion`

~~~suggestion
  ```
~~~

Just in case you come across this in future, it has bothered me for long before I knew about the `~~~`

Reviewer: That's awesome, thanks @Youssef1313 !

> 

Reviewer: ```suggestion
```

## When to suppress errors

> This is only recommended for advanced use cases where diagnostics on APIs need to explicitly disabled and the user is willing to take on the responsiblity of marking preview APIs appropriately. Disable the analyzer locally as shown in the code snippet below. For ex: Consider a case where an existing type implements a new preview interface. Since the entire type cannot be marked as preview (for backwards compatibility), the diagnostic around the type definition can be disabled locally. Further, the preview interface implementations need to be marked as preview by the user. Now, the existing type can be used as before, but calls to the new interface methods will get diagnostics. System.Private.CoreLib.csproj uses this technique to expose generic math features on Int32, Double, Decimal etc.

Reviewer: ```suggestion
Suppressing errors from this rule is only recommended for advanced use cases where diagnostics on APIs need to explicitly disabled. In this case, you must be willing to take on the responsibility of marking preview APIs appropriately. For example, consider a case where an existing type implements a new preview interface. Since the entire type cannot be marked as preview (for backwards compatibility), the diagnostic around the type definition can be disabled locally. Further, you need to mark the preview interface implementations as preview. Now, the existing type can be used as before, but calls to the new interface methods will get diagnostics. *System.Private.CoreLib.csproj* uses this technique to expose generic math features on numeric types such as `Int32`, `Double`, and `Decimal`.

The following images show how to disable the CA2252 analyzer locally.  
```

> ![CA2252 - Suppress Detect Preview Feature Diagnostic](media/ca2252-advanced-1.png)

> > ![CA2252 - Mark Interface Implementations Explicitly](media/ca2252-advanced-2.png)

Reviewer: ```suggestion

> ![CA2252 - Mark Interface Implementations Explicitly](media/ca2252-advanced-2.png)
```

> 

Reviewer: ```suggestion
```

## See also

> - [Preview Feature Design Document](https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md)

Reviewer: ```suggestion
- [Preview feature design document](https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md)
```
